vm/src/protocol/object.rs

```diff
@@ -411,4 +415,92 @@ impl PyObject {
             )))
         })
     }
+
+    pub fn get_item<K: DictKey + IntoPyObject + Clone>(
+        &self,
+        needle: K,
+        vm: &VirtualMachine,
+    ) -> PyResult {
+        if let Some(dict) = self.downcast_ref_if_exact::<PyDict>(vm) {
+            return dict.get_item(needle, vm);
+        }
+
+        let needle = needle.into_pyobject(vm);
+
+        if let Ok(mapping) = PyMapping::try_protocol(self, vm) {
+            mapping.subscript(&needle, vm)
+        } else {
+            // TODO: sequence protocol
+            match vm.get_special_method(self.to_owned(), "__getitem__")? {
+                Ok(special_method) => return special_method.invoke((needle,), vm),
+                Err(obj) => {
+                    if obj.class().issubclass(&vm.ctx.types.type_type) {
+                        if obj.is(&vm.ctx.types.type_type) {
+                            return PyGenericAlias::new(obj.clone_class(), needle, vm)
+                                .into_pyresult(vm);
+                        }
+
+                        if let Some(class_getitem) =
+                            vm.get_attribute_opt(obj, "__class_getitem__")?
+                        {
+                            return vm.invoke(&class_getitem, (needle,));
+                        }
+                    }
+                }
+            }
+            Err(vm.new_type_error(format!("'{}' object is not subscriptable", self.class())))
+        }
+    }
+
+    pub fn set_item<K: DictKey + IntoPyObject>(
+        &self,
+        needle: K,
+        value: PyObjectRef,
+        vm: &VirtualMachine,
+    ) -> PyResult<()> {
+        if let Some(dict) = self.downcast_ref_if_exact::<PyDict>(vm) {
+            return dict.set_item(needle, value, vm);
+        }
+
+        let needle = needle.into_pyobject(vm);
+
+        if let Ok(mapping) = PyMapping::try_protocol(self, vm) {
+            mapping.ass_subscript(&needle, Some(value), vm)
```

Comment:
According to [cpython code](https://github.com/python/cpython/blob/main/Objects/abstract.c#L208-L213), they check `ass_subscript` exist then call it, otherwise pass. For now, It seems we just check only mapping & `subscript` and call `ass_subscript`